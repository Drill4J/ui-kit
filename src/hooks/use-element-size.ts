/* eslint-disable @typescript-eslint/ban-ts-ignore */
import { useState, useEffect } from 'react';

export function useElementSize<E extends HTMLElement>(ref: React.RefObject<E>) {
  const [size, setSize] = useState({ width: 0, height: 0 });

  useEffect(() => {
    const element = ref && ref.current;
    function handleResize() {
      if (ref && ref.current) {
        const { width, height } = ref.current.getBoundingClientRect();
        setSize({ width, height });
      }
    }
    handleResize();
    // @ts-ignore
    const resizeObserver = new ResizeObserver(() => handleResize());
    resizeObserver.observe(element);

    return () => {
      resizeObserver.disconnect(element);
    };
  }, [ref]);

  return size;
}
