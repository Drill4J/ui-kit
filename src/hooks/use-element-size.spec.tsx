import { FC, useRef } from 'react';
import renderer from 'react-test-renderer';

import { useElementSize } from './use-element-size';

const Component: FC = () => {
  const node = useRef<HTMLDivElement>(null);
  const { width: contentWidth, height: contentHeight } = useElementSize(node);

  return (
    <div ref={node} style={{ display: 'inline-block' }}>
      <textarea placeholder={`width: ${contentWidth} and height: ${contentHeight}`} />
    </div>
  );
};

describe('useElementSize', () => {
  it('should match snapshot', () => {
    const component = renderer.create(<Component />).toJSON();

    expect(component).toMatchSnapshot();
  });
});
