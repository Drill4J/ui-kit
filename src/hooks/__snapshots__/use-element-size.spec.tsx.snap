// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useElementSize should match snapshot 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
    }
  }
>
  <textarea
    placeholder="width: 0 and height: 0"
  />
</div>
`;
